webapp/security/schemas.py

```diff
@@ -1,33 +1,97 @@
+        try:
+            date = dateutil.parser.parse(value)
+        except (OverflowError, ValueError):
+            raise self.make_error("parse_error", input=value)
+
+        return super()._deserialize(date.isoformat(), attr, data, **kwargs)
+
+
+# Schemas
+# ===
+
+
+# Notices
+# --
+def _validate_release_codenames(release_packages):
+    codenames = set(rel.codename for rel in db_session.query(Release).all())
+    unrecognised_codenames = set(release_packages.keys()) - codenames
+
+    if unrecognised_codenames:
+        raise ValidationError(
+            f"Unrecognised release codenames: {unrecognised_codenames}"
+        )
+
+
+class Package(Schema):
+    name = Str(required=True)
+    version = Str(required=True)
+    description = Str(required=True)
 
 
 class NoticeSchema(Schema):
-    notice_id = fields.Str(data_key="id", required=True)
-    title = fields.Str(required=True)
-    summary = fields.Str(required=True)
-    isummary = fields.Str()
-    description = fields.Str(required=True)
-    action = fields.Str()
-    releases = fields.Dict(required=True)
-    references = fields.List(fields.Str(), data_key="cves")
-    timestamp = fields.Float(required=True)
+    id = Str(required=True)
+    title = Str(required=True)
+    summary = Str(required=True)
+    instructions = Str(required=True)
+    references = List(Str())
+    cves = List(Str(validate=Regexp(r"(cve-|CVE-)\d{4}-\d{4,7}")))
+    published = ParsedDateTime(required=True)
+    description = Str(allow_none=True)
+    release_packages = Dict(
+        keys=Str(),
+        values=List(Nested(Package), required=True),
+        validate=_validate_release_codenames,
+    )
+
+
+# CVEs
+# --
+class Status(Schema):
+    release_codename = Str(required=True)
+    status = Str(required=True)
+    description = Str()
+
+
+class Package(Schema):
+    name = Str(required=True)
+    source = Str(required=True)
+    ubuntu = Str(required=True)
+    debian = Str(required=True)
+    statuses = List(Nested(Status))
+
+
+class Note(Schema):
+    author = Str(required=True)
+    note = Str(required=True)
 
 
 class CVESchema(Schema):
-    cve_id = fields.Str(data_key="id", required=True)
-    candidate = fields.Str(required=True)
-    public_date = fields.Str(required=True)
-    public_date_usn = fields.Str()
-    crd = fields.Str()
-    description = fields.Str()
-    ubuntu_description = fields.Str()
-    notes = fields.Str()
-    mitigation = fields.Str()
-    priority = fields.Str()
-    discovered_by = fields.Str()
-    assigned_to = fields.Str()
-    approved_by = fields.Str()
-    cvss = fields.Str()  # CVSS 3 and Base Score
-    packages = fields.List(fields.Str(), data_key="package")
-    references = fields.List(fields.Str(), data_key="cve_references")
-    bugs = fields.List(fields.Str(), data_key="cve_bugs")
+    id = Str(required=True)
+    published = ParsedDateTime()
+    description = Str()
+    ubuntu_description = Str()
+    notes = List(Nested(Note))
```

Comment:
If this has been changed into a list, do we not need to update the `webapp/security/models.py` in line 40 instead of Json, this would be a list?

Reply:
A list can be put in a JSON object perfectly fine.

I was using the ARRAY type, but this is PostgreSQL specific, and so less portable. JSON is more portable as it's supported in SQLite and MySQL too.